### PR TITLE
Refactoring Archive Viewer as hybrid PyDM display

### PR DESCRIPTION
**Pull request by zdomke**
### Refactored Archive Viewer as a hybrid PyDM Display
Changing the basis of the application from running on a pure python implementation, to running as a hybrid version. This should use both a .ui file and several .py files. The .ui file(s) will define the basic design of the application for This includes adding many new files as well as removing unused files.

Created files:
- archive_viewer/archive_viewer.py
  - Main script in the application. This starts the application and initializes the UI, making it functional.
  - This can be start up using `pydm archive_viewer.py` when run in the archive_viewer/ directory.
- archive_viewer/archive_viewer.ui
  - Created a PyDM UI for the application. Functionality is added in python files.
- archive_viewer/config.json
  - Added a config file for the application. This JSON is imported in the file archive_viewer/config.py .
- archive_viewer/config.py
  - One main configuration script that serializes the config JSON file as well as captures the application's logger. These can then be used throughout the application by importing the config module.
- archive_viewer/mixins/__init__.py
  - Make mixins/ importable
- archive_viewer/mixins/archivers_tab.py
  - Create the mixin for managing the Archivers Tab in the Properties section of the application.
  - The mixin class instantiates the table model and item delegate used by the table view in the Archivers Tab.
  - The mixin also adds functionality to the Add Archiver button in the Archivers Tab.
- archive_viewer/mixins/pv_table.py
  - Create the mixin for managing the Traces tab in the Properties section of the application, as well as defining 2 classes used by the PV Table's View.
  - The mixin class instantiates the table model, item delegate, and context menu used by the PV Table.
- archive_viewer/table_models/__init__.py
  - Make table_models/ importable
- archive_viewer/table_models/archiver_table_model.py
  - Defines the table model used by the table view in the Archivers Tab.
  - This model stores data on which archivers to fetch data from. It is initially populated with whichever archivers are included in the config JSON file, but allows users to add more archivers as desired.
  - The model checks the validity of archivers when enabled and/or their URLs are edited.
- archive_viewer/table_models/pv_table_model.py
  - Defines the table model that stores the data used by the PV Table in the Traces tab.
  - Checks the validity of PVs when they are added to the table.
  - Allows users to change color of lines representing PVs.
- archive_viewer/widgets/__init__.py
  - Make widgets/ importable
- archive_viewer/widgets/item_delegates.py
  - Define item delegates for the PV Table and Archivers Table.
  - The delegates populate the tables with widgets, allowing the user to interact with the data in these tables.
- archive_viewer/widgets/table_widgets.py
  - Define the ColorButton and CenterCheckbox widgets.
    - The ColorButton will open a color selector dialog window, allowing the user to select a new color for the button. Right clicking the button will return it to its default color.
    - The CenterCheckbox is a QWidget that contains a checkbox that will always be in the center. This is necessary as QTableViews are unable to contain a checkbox in the center of a cell.

Removed files:
- archive_viewer/main.py
  - Replaced by files archive_viewer/archive_viewer.py and archive_viewer/archive_viewer.ui
- archive_viewer/pv_table.py
  - The PV Table is now a Table View/Model, with the table view being defined in archive_viewer/archive_viewer.ui and the model being defined in archive_viewer/table_models/pv_table_model.py
- archive_viewer/pv_table_alt.py
  - Removed as it is redundant with archive_viewer/pv_table.py
- archive_viewer/range_axis_table.py
  - Removed for now
- archive_viewer/time_axis_table.py
  - Removed for now

Moved file:
- archive_viewer/archive_search.py  =>  archive_viewer/widgets/archive_search.py

### CODE REVIEW

File: archive_viewer/widgets/item_delegates.py [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+    Parameters
+    ----------
+    parent : QObject
+        The parent object for the DeleteRowDelegate.
+    table_model : QAbstractTableModel
+        The table model to be associated with the DeleteRowDelegate.
+    table_view : QTableView
+        The table view to be associated with the DeleteRowDelegate.
+    """
+    def __init__(self, parent: QObject, table_model: QAbstractTableModel, table_view: QTableView):
```
Comment (YektaY) [inline archive_viewer/widgets/item_delegates.py:244]:
```suggestion
    def __init__(self, parent: QObject, table_model: QAbstractTableModel, table_view: QTableView) -> None:
```

**Review by YektaY — COMMENTED** [review]
This looks really good to me. I have like forty comments, but like thirty-nine of them are on type hints. 

I also noticed a couple print statements left in the code, I skipped over them because I figured the project is still in development

Great work!

File: archive_viewer/table_models/archiver_table_model.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+        """Append an empty row to the end of the table model."""
+        new_row = [False, "", ""]
+        logger.debug("Adding row to archivers table.")
+
+        self.beginInsertRows(QModelIndex(), len(self._data), len(self._data))
+        self._data.append(new_row)
+        self._error.append(False)
+        self.endInsertRows()
+
+    def get_active_archivers(self) -> Dict[str, str]:
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/table_models/archiver_table_model.py:150]: Already contains return type hint.

File: archive_viewer/table_models/archiver_table_model.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+        """Return a dictionary of archivers that are enabled."""
+        return {n: u for a, n, u in self._data if a}
+
+    def get_all_archivers(self) -> Dict[str, str]:
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/table_models/archiver_table_model.py:154]: Already contains return type hint.

File: archive_viewer/table_models/axis_model.py [hunk 4]
```diff
@@ -0,0 +1,17 @@
+from pydm.widgets.axis_table_model import BasePlotAxesModel
+
+
+class ArchiverAxisModel(BasePlotAxesModel):
+    """The data model for the axes tab in the properties section. Acts
+    as a go-between for the axes in a plot, and QTableView items.
+    """
+    def append(self, name: str = ""):
+        """Add an empty row to the end of the table model."""
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/table_models/axis_model.py]: Added in commit 6d91981

File: archive_viewer/widgets/archive_search.py [hunk 5]
```diff
@@ -108,7 +100,7 @@ class ArchiveSearchWidget(QWidget):
     """
 
     def __init__(self, parent: Optional[QObject] = None):
-        super().__init__(parent=parent)
+        super(QWidget, self).__init__(parent=parent)
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/widgets/archive_search.py]: Fixed in [f9211b6]

File: archive_viewer/widgets/item_delegates.py (hunk 1 above)
Comment (zdomke) [reply archive_viewer/widgets/item_delegates.py]: Added in commit b1671bf

File: archive_viewer/widgets/item_delegates.py [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+    Parameters
+    ----------
+    parent : QObject
+        The parent object for the ColorButtonDelegate.
+    table_model : QAbstractTableModel
+        The table model to be associated with the ColorButtonDelegate.
+    table_view : QTableView
+        The table view to be associated with the ColorButtonDelegate.
+    """
+    def __init__(self, parent: QObject, table_model: QAbstractTableModel, table_view: QTableView):
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/widgets/item_delegates.py]: Added in commit b1671bf

File: archive_viewer/widgets/item_delegates.py [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+    Parameters
+    ----------
+    parent : QObject
+        The parent object for the CheckboxDelegate.
+    table_model : QAbstractTableModel
+        The table model to be associated with the CheckboxDelegate.
+    table_view : QTableView
+        The table view to be associated with the CheckboxDelegate.
+    """
+    def __init__(self, parent: QObject, table_model: QAbstractTableModel, table_view: QTableView):
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/widgets/item_delegates.py]: Added in commit b1671bf

File: archive_viewer/widgets/item_delegates.py [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+        return super().eventFilter(object, event)
+
+    @Slot(list)
+    def reset_items(self, items: list):
+        """"Clear the QComboBox's items and set them to the new list."""
+        self.editor.clear()
+        self.editor.addItems(items)
+
+    @Slot(QPoint)
+    def combo_menu_requested(self, pos: QPoint):
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/widgets/item_delegates.py]: Added in commit b1671bf

File: archive_viewer/widgets/item_delegates.py [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+        model.setData(index, data, Qt.EditRole)
+
+    def eventFilter(self, object: QObject, event: QEvent):
+        """Disable scrolling for widgets that are not the focus."""
+        if event.type() == QEvent.Wheel and not object.hasFocus():
+            return True
+        return super().eventFilter(object, event)
+
+    @Slot(list)
+    def reset_items(self, items: list):
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/widgets/item_delegates.py]: Added in commit b1671bf

File: archive_viewer/table_models/curve_model.py [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+    ----------
+    parent (optional) : QObject
+        The parent object for the table model.
+    plot : BasePlot
+        The plotting widget that the curves will be displayed on.
+    axis_model : ArchiverAxisModel
+        The table model that stores the axes for the plot.
+    """
+
+    def __init__(self, parent: Optional[QObject], plot: BasePlot, axis_model: ArchiverAxisModel):
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/table_models/curve_model.py]: Added in commit b1671bf

File: archive_viewer/mixins/traces_table.py [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+                            QVBoxLayout, QGridLayout, QLineEdit, QPushButton)
+from config import logger
+from widgets import (ArchiveSearchWidget, ColorButtonDelegate, CheckboxDelegate,
+                     ComboBoxDelegate, CurveStyleDelegate, DeleteRowDelegate)
+from table_models import ArchiverCurveModel
+
+
+class TracesTableMixin:
+    """Mixins class for the Traces tab of the settings section."""
+    def traces_table_init(self):
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/mixins/traces_table.py]: Added in commit b1671bf

File: archive_viewer/mixins/traces_table.py [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+        self.curves_model = ArchiverCurveModel(self, self.ui.archiver_plot, self.axis_table_model)
+        self.curves_model.append()
+
+        self.ui.traces_tbl.setModel(self.curves_model)
+
+        self.menu = PVContextMenu(self)
+        self.ui.traces_tbl.customContextMenuRequested.connect(
+            self.custom_context_menu)
+
+    def curve_delegates_init(self):
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/mixins/traces_table.py]: Added in commit b1671bf

File: archive_viewer/mixins/traces_table.py [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+    #   - Init: Loop through dict values:
+    #       - Create action w/ name
+    #       - action.triggered.connect(function)
+    #       - self.addAction(action)
+
+    # data_changed_signal = Signal(int)
+
+    # TODO: Archived PVs are no longer draggable from the search tool. Find out why
+
+    def __init__(self, parent: QObject = None):
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/mixins/traces_table.py]: Added in commit b1671bf

File: archive_viewer/widgets/table_widgets.py [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+
+        self.setLayout(layout)
+
+    @property
+    def checkState(self):
+        """Check state getter."""
+        return self.check_box.isChecked()
+
+    @checkState.setter
+    def checkState(self, state):
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/widgets/table_widgets.py]: Added in commit b1671bf

File: archive_viewer/widgets/table_widgets.py [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+
+    Parameters
+    ----------
+    parent : QObject
+        The parent object for the ComboBoxDelegate.
+    init_data : bool, optional
+        Whether the QCheckBox should be checked by default, by default True.
+    """
+    toggled = Signal(bool)
+    def __init__(self, parent: QObject, init_data: bool = True):
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/widgets/table_widgets.py]: Added in commit b1671bf

File: archive_viewer/widgets/table_widgets.py [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+    ----------
+    color : QColor or str, optional
+        Default color for the button to use, by default None
+    index : int, optional
+        A value used in determining a default color, by default -1
+    """
+
+    color_changed = Signal(QColor)
+
+    def __init__(self, *args: Any, color: Union[QColor, str] = None, index: int = -1, **kwargs):
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/widgets/table_widgets.py]: Added in commit b1671bf

File: archive_viewer/mixins/traces_table.py [hunk 17]
```diff
@@ -0,0 +1,224 @@
...
+        import_action = QAction("IMPORT CSV", self)
+        import_action.triggered.connect(self.import_csv)
+        self.addAction(import_action)
+
+    @property
+    def selected_index(self):
+        return self._selected_index
+
+    @selected_index.setter
+    def selected_index(self, ind: QModelIndex):
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/mixins/traces_table.py]: Added in commit b1671bf

File: archive_viewer/mixins/traces_table.py [hunk 18]
```diff
@@ -0,0 +1,224 @@
...
+        formula_action = QAction("FORMULA", self)
+        formula_action.triggered.connect(self._formula_dialog.exec_)
+        self.addAction(formula_action)
+
+        import_action = QAction("IMPORT CSV", self)
+        import_action.triggered.connect(self.import_csv)
+        self.addAction(import_action)
+
+    @property
+    def selected_index(self):
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/mixins/traces_table.py]: Added in commit b1671bf

File: archive_viewer/mixins/traces_table.py [hunk 19]
```diff
@@ -0,0 +1,224 @@
...
+        return self._selected_index
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/mixins/traces_table.py:143]: Added in commit 6d91981

File: archive_viewer/table_models/archiver_table_model.py [hunk 20]
```diff
@@ -0,0 +1,176 @@
+from typing import (Dict, Union)
+from qtpy.QtGui import QColor
+from qtpy.QtCore import (QModelIndex, Qt, Slot, QAbstractTableModel, QUrl)
+from qtpy.QtWidgets import QWidget
+from qtpy.QtNetwork import (QNetworkAccessManager, QNetworkReply, QNetworkRequest)
+from config import logger
+
+
+class ArchiversTableModel(QAbstractTableModel):
+    def __init__(self, parent: QWidget, init_archivers):
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/table_models/archiver_table_model.py]: Return type hint added in commit b1671bf

File: archive_viewer/table_models/archiver_table_model.py [hunk 21]
```diff
@@ -0,0 +1,176 @@
...
+        elif role == Qt.DisplayRole or role == Qt.EditRole:
+            return str(self._data[index.row()][index.column()])
+        elif role == Qt.UserRole:
+            return self._data[index.row()][index.column()]
+        elif role == Qt.ForegroundRole:
+            if self._error[index.row()]:
+                return QColor(Qt.red)
+            return QColor(Qt.black)
+
+    def setData(self, index: QModelIndex, value: Union[str, bool], role: Qt.ItemDataRole):
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/table_models/archiver_table_model.py]: Added in commit b1671bf

File: archive_viewer/mixins/traces_table.py [hunk 22]
```diff
@@ -0,0 +1,224 @@
...
+        self._selected_index = ind
+
+    @Slot()
+    def import_csv(self):
+        # TODO: Add action to import csv
+        pass
+
+
+class FormulaDialog(QDialog):
+    def __init__(self, parent: QObject):
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/mixins/traces_table.py]: Type hint added in commit b1671bf

File: archive_viewer/mixins/traces_table.py [hunk 23]
```diff
@@ -0,0 +1,224 @@
...
+        self.pv_name_input = QLineEdit(self)
+        self.pv_name_input.setPlaceholderText("Enter PV name")
+        layout.addWidget(self.pv_name_input)
+
+        # Add an "OK" button to accept the formula and close the dialog
+        ok_button = QPushButton("OK", self)
+        ok_button.clicked.connect(self.accept_formula)
+        layout.addWidget(ok_button)
+
+    def evaluate_formula(self, **kwargs: Dict[str, Any]):
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/mixins/traces_table.py]: Type hint added in commit b1671bf

File: archive_viewer/table_models/archiver_table_model.py [hunk 24]
```diff
@@ -0,0 +1,176 @@
...
+        self.network_manager = QNetworkAccessManager()
+        self.network_manager.finished.connect(self.archiver_validation)
+
+        self._headers = ["Fetch From", "Archiver", "URL"]
+
+        self._data = []
+        self._error = []
+        self.init_data(init_archivers)
+
+    def rowCount(self, index: QModelIndex = QModelIndex()):
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/table_models/archiver_table_model.py]: Type hint added in commit b1671bf

File: archive_viewer/table_models/archiver_table_model.py [hunk 25]
```diff
@@ -0,0 +1,176 @@
...
+        """Return the number of rows in the model."""
+        return len(self._data)
+
+    def columnCount(self, index: QModelIndex = QModelIndex()):
+        """Return the number of columns in the model."""
+        return len(self._headers)
+
+    def data(self, index: QModelIndex, role: Qt.ItemDataRole):
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/table_models/archiver_table_model.py]: Type hint added in commit b1671bf

File: archive_viewer/table_models/archiver_table_model.py [hunk 26]
```diff
@@ -0,0 +1,176 @@
...
+        if send_request:
+            reply = self.network_manager.get(QNetworkRequest(QUrl(url)))
+            reply.setProperty("index", enable_index)
+            logger.debug(f"Archiver validation request: {reply.url()}")
+
+        self._data[index.row()][index.column()] = value
+        self.dataChanged.emit(index, index)
+        return True
+
+    def headerData(self, section: int, orientation: Qt.Orientation, role: Qt.ItemDataRole):
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/table_models/archiver_table_model.py]: Type hint added in commit b1671bf

File: archive_viewer/table_models/archiver_table_model.py [hunk 27]
```diff
@@ -0,0 +1,176 @@
...
+        """Set the horizontal header's text."""
+        if orientation == Qt.Horizontal and role == Qt.DisplayRole:
+            return self._headers[section]
+
+    def flags(self, index: QModelIndex):
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/table_models/archiver_table_model.py]: Type hint added in commit b1671bf

File: archive_viewer/table_models/archiver_table_model.py [hunk 28]
```diff
@@ -0,0 +1,176 @@
...
+        """Returns a list of the set flags for the given index."""
+        fl = super().flags(index)
+        fl |= (Qt.ItemIsSelectable | Qt.ItemIsEnabled | Qt.ItemIsEditable)
+        return fl
+
+    def removeRows(self, row: int, count:int,  parent: QModelIndex = QModelIndex()):
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/table_models/archiver_table_model.py]: Type hint added in commit b1671bf

File: archive_viewer/table_models/archiver_table_model.py [hunk 29]
```diff
@@ -0,0 +1,176 @@
...
+        if count < 1 or (row + count) > len(self._data):
+            return False
+
+        self.beginRemoveRows(parent, row, row + count - 1)
+        del self._data[row:(row + count)]
+        self.endRemoveRows()
+
+        return True
+
+    def init_data(self, init_archivers: Dict[str, str]):
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/table_models/archiver_table_model.py]: Type hint added in commit b1671bf

File: archive_viewer/table_models/archiver_table_model.py [hunk 30]
```diff
@@ -0,0 +1,176 @@
...
+        self.beginInsertRows(QModelIndex(), 0, len(new_data))
+        self._data = new_data
+        self._error = [False] * len(new_data)
+        self.endInsertRows()
+
+        for row in range(len(self._data)):
+            ind = self.index(row, 0)
+            self.setData(ind, True, Qt.EditRole)
+
+    def add_empty_row(self):
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/table_models/archiver_table_model.py]: Type hint added in commit b1671bf

File: archive_viewer/table_models/archiver_table_model.py [hunk 31]
```diff
@@ -0,0 +1,176 @@
...
+    def get_active_archivers(self) -> Dict[str, str]:
+        """Return a dictionary of archivers that are enabled."""
+        return {n: u for a, n, u in self._data if a}
+
+    def get_all_archivers(self) -> Dict[str, str]:
+        """Return a dictionary of all archivers."""
+        return {n: u for _, n, u in self._data}
+
+    @Slot(QNetworkReply)
+    def archiver_validation(self, reply: QNetworkReply):
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/table_models/archiver_table_model.py]: Type hint added in commit b1671bf

File: archive_viewer/table_models/curve_model.py [hunk 32]
```diff
@@ -0,0 +1,169 @@
...
+        def_y_axis = self._axis_model.index(0, 0).data()
+        if not color:
+            color = ColorButton.index_color(self.rowCount())
+        self._row_names.append(self.next_header())
+        #          KLYS:LI22:31:KVAC
+        self.beginInsertRows(QModelIndex(), len(self._plot._curves), len(self._plot._curves))
+        self._plot.addYChannel(y_channel=address, name=name, color=color, useArchiveData=True, yAxisName=def_y_axis)
+        self.endInsertRows()
+
+    def removeAtIndex(self, index: QModelIndex):
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/table_models/curve_model.py]: Type hint added in commit b1671bf

File: archive_viewer/widgets/item_delegates.py [hunk 33]
```diff
@@ -0,0 +1,264 @@
...
+        The parent object for the SliderDelegate.
+    table_model : QAbstractTableModel
+        The table model to be associated with the SliderDelegate.
+    table_view : QTableView
+        The table view to be associated with the SliderDelegate.
+    init_range : Tuple[int, int], optional
+        The range for the QSlider widget, by default (1, 10)
+    """
+    def __init__(self, parent: Optional[QObject], table_model: QAbstractTableModel,
+                 table_view: QTableView, init_range: Tuple[int, int] = (1, 10)):
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/widgets/item_delegates.py]: Type hint added in commit b1671bf

File: archive_viewer/widgets/item_delegates.py [hunk 34]
```diff
@@ -0,0 +1,264 @@
...
+        The table model to be associated with the ComboBoxDelegate.
+    table_view : QTableView
+        The table view to be associated with the ComboBoxDelegate.
+    input : ArchiverAxisModel, list
+        The initial dataset to use when populating the QComboBox.
+    """
+    text_change_signal = Signal(int, str)
+
+    def __init__(self, parent: QObject, table_model: QAbstractTableModel,
+                 table_view: QTableView, input: Union[ArchiverAxisModel, list]):
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/widgets/item_delegates.py]: Type hint added in commit b1671bf

File: archive_viewer/widgets/item_delegates.py [hunk 35]
```diff
@@ -0,0 +1,264 @@
...
+        editor.setCurrentText(value)
+
+    def setModelData(self, editor: QComboBox, model: QAbstractTableModel, index: QModelIndex) -> None:
+        """Set the table model's data to match the editor's data."""
+        data = editor.currentIndex()
+        new_axis_name = editor.currentText()
+        self.text_change_signal.emit(index.row(), new_axis_name)
+        model.setData(index, data, Qt.EditRole)
+
+    def eventFilter(self, object: QObject, event: QEvent):
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/widgets/item_delegates.py]: Type hint added in commit b1671bf

File: archive_viewer/widgets/table_widgets.py [hunk 36]
```diff
@@ -0,0 +1,129 @@
...
+        black background."""
+        h = int(360 * random())
+        s = int(256 * (0.5 + random()/2.0))
+        l = int(256 * (0.4 + random()/5.0))
+        color = QColor()
+        color.setHsl(h, s, l)
+        return color
+
+    @staticmethod
+    def index_color(index: int):
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/widgets/table_widgets.py]: Type hint added in commit b1671bf

File: archive_viewer/mixins/traces_table.py [hunk 37]
```diff
@@ -0,0 +1,224 @@
...
+        #   if yes, replace with row header; if no, add to TableModel and replace with row header
+        formula = self.field.text()
+        try:
+            result = eval(formula)
+            self.field.setText(str(result))
+        except (SyntaxError, TypeError):
+            self.field.setText("Error")
+            logger.error("Invalid formula evaluated.")
+
+    def accept_formula(self, **kwargs: Dict[str, Any]):
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/mixins/traces_table.py]: Type hint added in commit b1671bf

File: archive_viewer/table_models/archiver_table_model.py [hunk 38]
```diff
@@ -0,0 +1,176 @@
...
+
+        self._data = []
+        self._error = []
+        self.init_data(init_archivers)
+
+    def rowCount(self, index: QModelIndex = QModelIndex()):
+        """Return the number of rows in the model."""
+        return len(self._data)
+
+    def columnCount(self, index: QModelIndex = QModelIndex()):
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/table_models/archiver_table_model.py]: Type hint added in commit b1671bf

File: archive_viewer/table_models/archiver_table_model.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/table_models/archiver_table_model.py]: init_archivers type hint added in commit 68fbc3d

File: archive_viewer/table_models/axis_model.py [hunk 39]
```diff
@@ -0,0 +1,17 @@
+from pydm.widgets.axis_table_model import BasePlotAxesModel
+
+
+class ArchiverAxisModel(BasePlotAxesModel):
+    """The data model for the axes tab in the properties section. Acts
+    as a go-between for the axes in a plot, and QTableView items.
+    """
+    def append(self, name: str = ""):
```
[the remark this answers is not in the document]
Comment (zdomke) [reply archive_viewer/table_models/axis_model.py]: Return type hint added in commit b1671bf

**Pull request merged**